crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -127,14 +126,227 @@ pub trait QuiltDecoderApi<'a, V: QuiltVersion> {
     fn get_or_decode_quilt_index(&mut self) -> Result<&V::QuiltIndex, QuiltError>;
 
     /// Gets a blob by its identifier from the quilt.
-    fn get_blob_by_identifier(&self, identifier: &str) -> Result<Vec<u8>, QuiltError>;
+    fn get_blob_by_identifier(&self, identifier: &str) -> Result<QuiltStoreBlobOwned, QuiltError>;
 
     /// Adds slivers to the decoder.
     fn add_slivers(&mut self, slivers: &'a [&'a SliverData<Secondary>]);
 }
 
+/// A unified trait for iterating over column-based quilt data from different sources.
+pub trait QuiltColumnDataIterator {
+    type RangeIter<'a>: Iterator<Item = &'a [u8]> + 'a
+    where
+        Self: 'a;
+
+    /// Returns an iterator over bytes in the columns identified by the given start and end indices.
+    ///
+    /// Assuming a 0-indexed bytes, the iterator will return the bytes in the range
+    /// `[start_byte, start_byte + limit_bytes)`.
+    fn iter_column_bytes(
```

Comment:
Question: Do we ever use this for anything besides combining it into an output vector? If not, we could probably remove and/or simplify some of the code. See also my comment about a `collect_to_vec` function above.